### PR TITLE
update dashboard_data signals

### PR DESCRIPTION
**Pull request by p-galligan**
Change `pre_delete` to `post_delete`.

### CODE REVIEW

**Comment by p-galligan** [comment]
@helrond I'm struggling a bit with this one.

Tests are failing due to line 135 of `helpers.py`. The function is:

`def delete_test_orgs(orgs=[]):
    for org in orgs:
        org.delete()`

I'm struggling on understanding why what I did is changing this code. Any ideas?

**Comment by helrond** [comment]
@p-galligan Can you paste a stack trace or a specific error message?

**Comment by p-galligan** [comment]
Super long, but here's the first error that pops up in the tests.

```
##########################################
TRANSFERS.DELIVER_TRANSFERS CRON END
2020-05-27 12:07:27.661078
##########################################


ETest organization Trump A Group -- trumpagroup created
ETest organization Clinton g Group -- clintonggroup created
Test organization Trump Z Org -- trumpzorg created
Test organization Trump M Foundation -- trumpmfoundation created
No active organizations in database
1 orgs were removed from routine
.
======================================================================
ERROR: test_accessioning (bag_transfer.test.test_accessioning.AccessioningTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/mysql/base.py", line 71, in execute
    return self.cursor.execute(query, args)
  File "/usr/local/lib/python3.6/site-packages/MySQLdb/cursors.py", line 209, in execute
    res = self._query(query)
  File "/usr/local/lib/python3.6/site-packages/MySQLdb/cursors.py", line 315, in _query
    db.query(q)
  File "/usr/local/lib/python3.6/site-packages/MySQLdb/connections.py", line 239, in query
    _mysql.connection.query(self, query)
MySQLdb._exceptions.IntegrityError: (1451, 'Cannot delete or update a parent row: a foreign key constraint fails (`test_mysql`.`bag_transfer_dashboardmonthdata`, CONSTRAINT `bag_transfer_dashboa_organization_id_f5a0444d_fk_bag_trans` FOREIGN KEY (`organization_id`) REFERENCES `bag_transfer_organization` ()')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/code/aurora/bag_transfer/test/test_accessioning.py", line 89, in tearDown
    helpers.delete_test_orgs(self.orgs)
  File "/code/aurora/bag_transfer/test/helpers.py", line 135, in delete_test_orgs
    org.delete()
  File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 919, in delete
    return collector.delete()
  File "/usr/local/lib/python3.6/site-packages/django/db/models/deletion.py", line 312, in delete
    count = query.delete_batch(pk_list, self.using)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/subqueries.py", line 41, in delete_batch
    num_deleted += self.do_query(self.get_meta().db_table, self.where, using=using)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/subqueries.py", line 24, in do_query
    cursor = self.get_compiler(using).execute_sql(CURSOR)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1133, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 76, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/utils.py", line 89, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/mysql/base.py", line 71, in execute
    return self.cursor.execute(query, args)
  File "/usr/local/lib/python3.6/site-packages/MySQLdb/cursors.py", line 209, in execute
    res = self._query(query)
  File "/usr/local/lib/python3.6/site-packages/MySQLdb/cursors.py", line 315, in _query
    db.query(q)
  File "/usr/local/lib/python3.6/site-packages/MySQLdb/connections.py", line 239, in query
    _mysql.connection.query(self, query)
django.db.utils.IntegrityError: (1451, 'Cannot delete or update a parent row: a foreign key constraint fails (`test_mysql`.`bag_transfer_dashboardmonthdata`, CONSTRAINT `bag_transfer_dashboa_organization_id_f5a0444d_fk_bag_trans` FOREIGN KEY (`organization_id`) REFERENCES `bag_transfer_organization` ()')

======================================================================
ERROR: test_api (bag_transfer.test.test_api.APITest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/mysql/base.py", line 71, in execute
    return self.cursor.execute(query, args)
  File "/usr/local/lib/python3.6/site-packages/MySQLdb/cursors.py", line 209, in execute
    res = self._query(query)
  File "/usr/local/lib/python3.6/site-packages/MySQLdb/cursors.py", line 315, in _query
    db.query(q)
  File "/usr/local/lib/python3.6/site-packages/MySQLdb/connections.py", line 239, in query
    _mysql.connection.query(self, query)
MySQLdb._exceptions.IntegrityError: (1451, 'Cannot delete or update a parent row: a foreign key constraint fails (`test_mysql`.`bag_transfer_dashboardmonthdata`, CONSTRAINT `bag_transfer_dashboa_organization_id_f5a0444d_fk_bag_trans` FOREIGN KEY (`organization_id`) REFERENCES `bag_transfer_organization` ()')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/code/aurora/bag_transfer/test/test_api.py", line 108, in tearDown
    helpers.delete_test_orgs(self.orgs)
  File "/code/aurora/bag_transfer/test/helpers.py", line 135, in delete_test_orgs
    org.delete()
  File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 919, in delete
    return collector.delete()
  File "/usr/local/lib/python3.6/site-packages/django/db/models/deletion.py", line 312, in delete
    count = query.delete_batch(pk_list, self.using)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/subqueries.py", line 41, in delete_batch
    num_deleted += self.do_query(self.get_meta().db_table, self.where, using=using)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/subqueries.py", line 24, in do_query
    cursor = self.get_compiler(using).execute_sql(CURSOR)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1133, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 76, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/utils.py", line 89, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/mysql/base.py", line 71, in execute
    return self.cursor.execute(query, args)
  File "/usr/local/lib/python3.6/site-packages/MySQLdb/cursors.py", line 209, in execute
    res = self._query(query)
  File "/usr/local/lib/python3.6/site-packages/MySQLdb/cursors.py", line 315, in _query
    db.query(q)
  File "/usr/local/lib/python3.6/site-packages/MySQLdb/connections.py", line 239, in query
    _mysql.connection.query(self, query)
django.db.utils.IntegrityError: (1451, 'Cannot delete or update a parent row: a foreign key constraint fails (`test_mysql`.`bag_transfer_dashboardmonthdata`, CONSTRAINT `bag_transfer_dashboa_organization_id_f5a0444d_fk_bag_trans` FOREIGN KEY (`organization_id`) REFERENCES `bag_transfer_organization` ()')
````

**Comment by helrond** [comment]
OK, that's super helpful.

What's happening here is that the DashboardMonthData objects that you've now created (correctly) have references to those organization objects, so you can't just delete the organizations because it would make the DashboardMonthData objects invalid. I think what this means is that we need to add some behavior to the Organization `pre_delete` signal to delete all associated DashboardMonthData objects tied to the organization.

**Comment by helrond** [comment]
I'm also looking at the `on_delete` setting for DashboardMonthDatas and I think it should _NOT_ be `models.CASCADE` because that will delete the organization, which we do not want.

**Comment by p-galligan** [comment]
@helrond This is mostly working, but I can't figure out why I have a 1-record disconnect between the side tab and bottom tab statistics

File: aurora/bag_transfer/signals.py [hunk 1]
```diff
@@ -25,10 +26,9 @@ def set_is_staff(sender, instance, **kwargs):
 
 
 @receiver(post_save, sender=Archives)
-@receiver(pre_delete, sender=Archives)
 def dashboard_data(sender, instance, **kwargs):
```
Comment (helrond) [inline aurora/bag_transfer/signals.py]:
```suggestion
def add_dashboard_data(sender, instance, **kwargs):
```

File: aurora/bag_transfer/signals.py [hunk 2]
```diff
@@ -42,33 +42,85 @@ def dashboard_data(sender, instance, **kwargs):
...
                     organization=organization,
                 )[0]
-                data.upload_count = Archives.objects.filter(
+                set_uploads(current, data, organization)
+            current += relativedelta(months=1)
+    set_count(sender, instance, organization)
+
+
+@receiver(post_delete, sender=Archives)
+def dashboard_check(sender, instance, **kwargs):
```
Comment (helrond) [inline aurora/bag_transfer/signals.py]:
```suggestion
def remove_dashboard_data(sender, instance, **kwargs):
```

File: aurora/bag_transfer/signals.py [hunk 3]
```diff
@@ -42,33 +42,85 @@ def dashboard_data(sender, instance, **kwargs):
...
+    """
+    Updates dashboard data each time a transfer is deleted, which
```
Comment (helrond) [inline aurora/bag_transfer/signals.py]:
```suggestion
    Removes dashboard data each time a transfer is deleted, which
```

File: aurora/bag_transfer/signals.py [hunk 4]
```diff
@@ -25,10 +26,9 @@ def set_is_staff(sender, instance, **kwargs):
 
 
 @receiver(post_save, sender=Archives)
-@receiver(pre_delete, sender=Archives)
 def dashboard_data(sender, instance, **kwargs):
     """
-    Updates dashboard data each time a transfer is saved or deleted, which
+    Updates dashboard data each time a transfer is saved, which
```
Comment (helrond) [inline aurora/bag_transfer/signals.py]:
```suggestion
    Adds dashboard data each time a transfer is saved, which
```

File: aurora/bag_transfer/signals.py [hunk 5]
```diff
@@ -42,33 +42,85 @@ def dashboard_data(sender, instance, **kwargs):
...
+    """
+    Updates dashboard data counts.
+    """
     if instance.process_status >= sender.VALIDATED:
         for organization in Organization.objects.all():
             for label in set(
                 BagInfoMetadata.objects.all().values_list("record_type", flat=True)
             ):
-                data = DashboardRecordTypeData.objects.get_or_create(
+                if DashboardRecordTypeData.objects.filter(
```
Comment (helrond) [inline aurora/bag_transfer/signals.py]: Do we need to use this logic here? In the `else` statement you're using `get_or_create` which means that a record type will ALWAYS be created. So can we just use `get_or_create`?

**Review by helrond — CHANGES_REQUESTED** [review]

File: aurora/bag_transfer/signals.py (hunk 5 above)
Comment (p-galligan) [reply aurora/bag_transfer/signals.py]: You're right, that wasn't necessary.

**Review by helrond — APPROVED** [review]

**Review by bonniegee — APPROVED** [review]

**Pull request merged**